alibi_detect/saving/_pytorch/loading.py

```diff
@@ -0,0 +1,143 @@
+    """
+    Load PyTorch model.
+
+    Parameters
+    ----------
+    filepath
+        Saved model filepath.
+    layer
+        Optional index of a hidden layer to extract. If not `None`, a
+        :py:class:`~alibi_detect.cd.pytorch.HiddenOutput` model is returned.
+
+    Returns
+    -------
+    Loaded model.
+    """
+    filepath = Path(filepath).joinpath('model.pt')
+    model = torch.load(filepath, pickle_module=dill)
+    # Optionally extract hidden layer
+    if isinstance(layer, int):
+        model = HiddenOutput(model, layer=layer)
+    return model
+
+
+def prep_model_and_emb(model: nn.Module, emb: Optional[TransformerEmbedding]) -> nn.Module:
+    """
+    Function to perform final preprocessing of model (and/or embedding) before it is passed to preprocess_drift.
+
+    Parameters
+    ----------
+    model
+        A compatible model.
+    emb
+        An optional text embedding model.
+
+    Returns
+    -------
+    The final model ready to passed to preprocess_drift.
+    """
+    # Process model (and embedding)
+    model = model.encoder if isinstance(model, UAE) else model  # This is to avoid nesting UAE's already a UAE
+    if emb is not None:
+        model = _Encoder(emb, mlp=model)
+        model = UAE(encoder_net=model)
+    return model
+
+
+def load_kernel_config(cfg: dict) -> Callable:
+    """
+    Loads a kernel from a kernel config dict.
+
+    Parameters
+    ----------
+    cfg
+        A kernel config dict. (see pydantic schema's).
+
+    Returns
+    -------
+    The kernel.
+    """
+    if 'src' in cfg:  # Standard kernel config
+        kernel = cfg.pop('src')
+        if hasattr(kernel, 'from_config'):
+            kernel = kernel.from_config(cfg)
+
+    elif 'proj' in cfg:  # DeepKernel config
+        # Kernel a
+        kernel_a = cfg['kernel_a']
+        kernel_b = cfg['kernel_b']
+        if kernel_a != 'rbf':
+            cfg['kernel_a'] = load_kernel_config(kernel_a)
+        if kernel_b != 'rbf':
+            cfg['kernel_b'] = load_kernel_config(kernel_b)
+        # Assemble deep kernel
+        kernel = DeepKernel.from_config(cfg)
+
+    else:
+        raise ValueError('Unable to process kernel. The kernel config dict must either be a `KernelConfig` with a '
+                         '`src` field, or a `DeepkernelConfig` with a `proj` field.)')
+    return kernel
+
+
+def load_optimizer(cfg: dict) -> Type[torch.optim.Optimizer]:
+    """
+    Imports a PyTorch torch.optim.Optimizer class from an optimizer config dict.
+
+    Parameters
+    ----------
+    cfg
+        The optimizer config dict.
+
+    Returns
+    -------
+    The loaded optimizer class.
+    """
+    class_name = cfg.get('class_name', None)
+    try:
+        return getattr(import_module('torch.optim'), class_name)
+    except AttributeError:
+        raise ValueError(f"{class_name} is not a recognised optimizer in `torch.optim`.")
+
+
+def load_embedding(src: str, embedding_type, layers) -> TransformerEmbedding:
+    """
```

Comment:
Pretty much same as tensorflow equivalent.

Reply:
Types of `embedding_type` and `layers`.

Reply:
What is the benefit of `load_embedding` if it just calls `TransformerEmbedding` with the same args/kwargs? Consistency in naming?

Reply:
Yeh just for consistency in the top-level (backend agnostic) `load_embedding_config`:

https://github.com/SeldonIO/alibi-detect/blob/ac01d352b31607473970307080096a9c7b4d5744/alibi_detect/saving/loading.py#L292-L300

Reply:
Actually sorry, that's not quite true. It's also hides the below import away in the `_pytorch.saving` submodule:

https://github.com/SeldonIO/alibi-detect/blob/ac01d352b31607473970307080096a9c7b4d5744/alibi_detect/saving/_pytorch/loading.py#L13

Hiding the backend-specific stuff in backend-specific submodules is something we've been generally trying to follow since the optional deps work (although admittedly here it leads to a bit of a needless abstraction...)